lib/axiom/fs/stream/memory_stream_buffer.js

```diff
@@ -24,114 +24,33 @@ var ReadableStream;
 /** @typedef WritableStream$$module$axiom$fs$stream$writable_stream */
 var WritableStream;
 
-/** @typedef function():void */
-var EventCallback;
-
-/**
- * Event value + associated callback
- * @constructor
- * @param {!*} value
- * @param {!EventCallback} callback
- */
-var EventWithCallback = function(value, callback) {
-  this.value = value;
-  this.callback = callback;
-};
-
 /**
- * @constructor
+ * @constructor @extends {StreamBuffer}
  */
 export var MemoryStreamBuffer = function() {
-  /** @private @type {boolean} */
-  this.paused_ = true;
-  /** @const @private @type {Queue} */
-  this.events_ = new Queue();
-  /** @type {!AxiomEvent} */
-  this.onData = new AxiomEvent();
-  /** @type {!AxiomEvent} */
-  this.onReadable = new AxiomEvent();
+  StreamBuffer.call(this);
+
   /** @const @type {!ReadableStream} */
   this.readableStream = new ReadableMemoryStream(this);
   /** @const @type {!WritableStream} */
   this.writableStream = new WritableMemoryStream(this);
+  
+  this.onClose = AxiomEvent.all([
```

Comment:
I think I see what you are doing here, but given my previous comments, I think it should be something along the lines of
* add an `end` method
* when `end` is called, flush all write events, then call the `onEnd' event on the readable stream and the `onFinish` event on the writable stream.